add monitoring content (#337)
digital-asset/docs.daml.com: docs/2.7.0/docs/canton/usermanual/example_monitoring_setup.rst

Container Setup
---------------

Following are the details for how you can configure `Docker Compose <https://docs.docker.com/compose/>`_ to spin up the Docker container network shown in the diagram. See the `compose` documentation for detailed information concerning the structure of the configuration files.

One feature of `compose` is that you can provide the overall configuration across multiple files. Each configuration file is described below, followed by information on how to bring them together in a running network.

.. image:: ./images/basic-canton-setup.svg
   :align: center
   :width: 100%
   :alt: A diagram showing an example Docker network setup

Intended Use
~~~~~~~~~~~~

This example is intended to demonstrate how you can expose, aggregate, and observe monitoring information from Canton. It is not suitable for production without alterations. Note the following warnings.

 .. warning::

>    Ports are exposed from the Docker network that are not necessary to support the UI. This may be to allow low-level interaction with the underlying service via a REST or similar interface. In a production system, the only ports that should be exposed are those required for the operation of the system.

Author: ```suggestion
   Ports are exposed from the Docker network that are not necessary to support the UI. For example, it can allow low-level interaction with the underlying service via a REST or similar interface. In a production system, the only ports that should be exposed are those required for the operation of the system.
```

>    Ports are exposed from the Docker network that are not necessary to support the UI. This may be to allow low-level interaction with the underlying service via a REST or similar interface. In a production system, the only ports that should be exposed are those required for the operation of the system.

Reviewer: > 

How about "the network can" rather than "it can"?

Dependencies
~~~~~~~~~~~~

There are startup dependencies between the Docker containers. For example, the domain needs to be running before the participant, and the database needs to run before the domain.

> The `yaml` anchor `x-logging` enabled GELF container logging and is duplicated across the containers where you want to capture logging output. Note that the host address is the host machine, not a network address (on OSX).

Author: @carrielaben-da @hrischuk-da I had trouble making sense of the sentence beginning with "The 'yaml' anchor". Hoping someone can help me edit it?

Reviewer 2: Highly technical gibberish.   Suggest accepting it

Author: Thanks, @hrischuk-da! I wasn't sure whether there was or wasn't a verb in the sentence. I'll leave it be.

- Good defaults for these products allow a basic setup to be started with almost zero configuration.
- The ELK setup acts as a good baseline as compared to other options such as `Loki`_ or `Graylog`_.

Use Prometheus/Grafana
~~~~~~~~~~~~~~~~~~~~~~

**Reasons:**

- Prometheus defines and uses the `OpenTelemetry`_ reference file format.
- Exposing metrics via an HTTP endpoint allows easy direct inspection of metric values.
- The Prometheus approach of pulling metrics from underlying system means that the running containers do not need infrastructure to store and push metric data.
- Grafana works very well with Prometheus.

Logging Improvements
--------------------
This version of the example only has the logging structure provided via GELF. It is possible to improve this by:

  - Extracting data from the underlying containers as a JSON stream.
  - Mapping fields in this JSON data onto the `ECS`_ so that the same name is used for commonly used field values (for example, log level).
  - Configuring Elasticsearch with a schema that allows certain fields to be quickly filtered (for example, log level).

Glossary
--------

Docker Log Driver
~~~~~~~~~~~~~~~~~

> Docker containers can be configured with a log driver that allows log output to be exported out of the Docker container. Using log drivers to export logging information means that it is not necessary to run another process on the Docker container to do this.

Reviewer: ```suggestion
Docker containers can be configured with a log driver that allows log output to be exported from the Docker container. Using log drivers to export logging information makes running another process on the Docker container for this unnecessary.
```
